File: raincoat-fabric/src/main/java/dev/uten2c/raincoat/model/GunUnbakedModel.java
package dev.uten2c.raincoat.model;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.collect.Lists;
import dev.uten2c.raincoat.util.StackUtils;
import net.minecraft.client.render.model.BakedModel;
import net.minecraft.client.render.model.Baker;
import net.minecraft.client.render.model.ModelBakeSettings;
import net.minecraft.client.render.model.UnbakedModel;
import net.minecraft.client.render.model.json.JsonUnbakedModel;
import net.minecraft.client.render.model.json.ModelElement;
import net.minecraft.client.texture.MissingSprite;
import net.minecraft.client.texture.Sprite;
import net.minecraft.client.util.SpriteIdentifier;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.PlayerScreenHandler;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Collections;
import java.util.function.Function;

public record GunUnbakedModel(@NotNull GunModel gunModel) implements UnbakedModel {
    private static final SpriteIdentifier MISSING_SPRITE_ID = new SpriteIdentifier(PlayerScreenHandler.BLOCK_ATLAS_TEXTURE, MissingSprite.getMissingSpriteId());
    private static final Cache<Integer, BakedModel> MODEL_CACHE = CacheBuilder.newBuilder()
            .weakKeys()
            .weakValues()
            .build();
    private static final Baker DUMMY_BAKER = new Baker() {
        @Override
        public UnbakedModel getOrLoadModel(Identifier id) {
            throw new AssertionError();
        }

        @Nullable
        @Override
        public BakedModel bake(Identifier id, ModelBakeSettings settings) {
            throw new AssertionError();
        }
    };

    @Override
    public Collection<Identifier> getModelDependencies() {
        return Collections.emptyList();
    }

    @Override
    public void setParents(Function<Identifier, UnbakedModel> modelLoader) {
    }

    @Override
    public @NotNull BakedModel bake(Baker baker, Function<SpriteIdentifier, Sprite> textureGetter, ModelBakeSettings rotationContainer, Identifier modelId) {
        return new GunBakedModel(textureGetter.apply(MISSING_SPRITE_ID), gunModel, textureGetter, rotationContainer, modelId);
    }

    public static @Nullable BakedModel getOrCreateModel(@NotNull ItemStack stack, @NotNull GunBakedModel model) {
        if (stack.isEmpty()) {
            return null;
        }
        final var gunModelState = StackUtils.getGunModelState(stack);
        if (gunModelState == -1) {
            return null;
        }
        final var cache = MODEL_CACHE.getIfPresent(gunModelState);
        if (cache != null) {
            return cache;
        }
        final var newModel = createBakedModel(model.gunModel(), gunModelState, model.textureGetter(), model.rotationContainer(), model.modelId());
        if (newModel == null) {
            return null;
        }
        MODEL_CACHE.put(gunModelState, newModel);
        return newModel;
    }

    private static @Nullable BakedModel createBakedModel(@NotNull GunModel gunModel, int gunModelState, @NotNull Function<SpriteIdentifier, Sprite> textureGetter, @NotNull ModelBakeSettings rotationContainer, @NotNull Identifier modelId) {
        final var removeIndices = gunModel.removeIndices().get(gunModelState);
        if (removeIndices == null) {
            return null;
        }
        final var modelTransformation = gunModel.modelTransformations().get(gunModelState);
        if (modelTransformation == null) {
            return null;
        }
        final var trimmedElements = Lists.<ModelElement>newArrayList();
        for (int i = 0; i < gunModel.elements().size(); i++) {
            if (removeIndices.contains(i)) {
                continue;
            }
            trimmedElements.add(gunModel.elements().get(i));
        }
        final var jsonUnbakedModel = new JsonUnbakedModel(null, trimmedElements, gunModel.textures(), true, JsonUnbakedModel.GuiLight.ITEM, modelTransformation, Collections.emptyList());
        return jsonUnbakedModel.bake(DUMMY_BAKER, textureGetter, rotationContainer, modelId);
    }

    public static void clearCache() {
        MODEL_CACHE.cleanUp();
    }
}

File: raincoat-fabric/src/main/kotlin/dev/uten2c/raincoat/model/GunUnbakedModel.kt
package dev.uten2c.raincoat.model

import com.google.common.cache.CacheBuilder
import com.google.common.collect.Lists
import dev.uten2c.raincoat.util.StackUtils
import net.minecraft.client.render.model.BakedModel
import net.minecraft.client.render.model.Baker
import net.minecraft.client.render.model.ModelBakeSettings
import net.minecraft.client.render.model.UnbakedModel
import net.minecraft.client.render.model.json.JsonUnbakedModel
import net.minecraft.client.render.model.json.ModelElement
import net.minecraft.client.render.model.json.ModelTransformation
import net.minecraft.client.texture.MissingSprite
import net.minecraft.client.texture.Sprite
import net.minecraft.client.util.SpriteIdentifier
import net.minecraft.item.ItemStack
import net.minecraft.screen.PlayerScreenHandler
import net.minecraft.util.Identifier
import java.util.function.Function

class GunUnbakedModel(private val gunModel: GunModel) : UnbakedModel {
    override fun getModelDependencies(): MutableCollection<Identifier> {
        return mutableListOf()
    }

    override fun setParents(modelLoader: Function<Identifier, UnbakedModel>?) {
    }

    override fun bake(baker: Baker, textureGetter: Function<SpriteIdentifier, Sprite>, rotationContainer: ModelBakeSettings, modelId: Identifier): BakedModel {
        return GunBakedModel(textureGetter.apply(missingSpriteId), gunModel, textureGetter, rotationContainer, modelId)
    }

    companion object {
        private val missingSpriteId = SpriteIdentifier(PlayerScreenHandler.BLOCK_ATLAS_TEXTURE, MissingSprite.getMissingSpriteId())
        private val modelCache = CacheBuilder.newBuilder()
            .weakKeys()
            .weakValues()
            .build<Int, BakedModel>()
        private val dummyBaker = object : Baker {
            override fun getOrLoadModel(id: Identifier): UnbakedModel {
                throw AssertionError()
            }

            override fun bake(id: Identifier, settings: ModelBakeSettings): BakedModel? {
                throw AssertionError()
            }
        }

        @JvmStatic
        fun getOrCreateModel(stack: ItemStack, model: GunBakedModel): BakedModel? {
            if (stack.isEmpty) {
                return null
            }
            val gunModelState = StackUtils.getGunModelState(stack)
            if (gunModelState == -1) {
                return null
            }
            val cache = modelCache.getIfPresent(gunModelState)
            if (cache != null) {
                return cache
            }
            val bakedModel = createBakedModel(model.gunModel, gunModelState, model.textureGetter, model.rotationContainer, model.modelId) ?: return null
            modelCache.put(gunModelState, bakedModel)
            return bakedModel
        }

        private fun createBakedModel(gunModel: GunModel, gunModelState: Int, textureGetter: Function<SpriteIdentifier, Sprite>, rotationContainer: ModelBakeSettings, modelId: Identifier): BakedModel? {
            val removeIndices = gunModel.removeIndices[gunModelState] ?: emptyList()
            val modelTransformation = gunModel.modelTransformations[gunModelState]?.toMinecraft() ?: ModelTransformation.NONE
            val trimmedElements = Lists.newArrayList<ModelElement>()
            for (i in gunModel.elements.indices) {
                if (removeIndices.contains(i)) {
                    continue
                }
                trimmedElements.add(gunModel.elements[i].toMinecraft())
            }
            val textures = gunModel.textures
                .mapValues { it.value.toMinecraft() }
            val jsonUnbakedModel = JsonUnbakedModel(null, trimmedElements, textures, true, JsonUnbakedModel.GuiLight.ITEM, modelTransformation, emptyList())
            return jsonUnbakedModel.bake(dummyBaker, textureGetter, rotationContainer, modelId)
        }

        fun clearCache() {
            modelCache.cleanUp()
        }
    }
}
